src/concrete/ml/sklearn/base.py

```diff
@@ -1690,3 +1695,385 @@ def predict_proba(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) ->
+
+        self._y = y
+
+        # Fit the scikit-learn model
+        self._fit_sklearn_model(X, y, **fit_parameters)
+
+        # Check that the underlying sklearn model has been set and fit
+        assert self.sklearn_model is not None, self._sklearn_model_is_not_fitted_error_message()
+
+        # Retrieve the ONNX graph
+        self._set_onnx_model(X)
+
+        # Convert the n_bits attribute into a proper dictionary
+        n_bits = get_n_bits_dict(self.n_bits)
+
+        input_n_bits = n_bits["op_inputs"]
+        input_options = QuantizationOptions(n_bits=input_n_bits, is_signed=True)
+
+        # Quantize the inputs and store the associated quantizer
+        q_inputs = QuantizedArray(n_bits=input_n_bits, values=X, options=input_options)
+        input_quantizer = q_inputs.quantizer
+        self.input_quantizers.append(input_quantizer)
+
+        weights_n_bits = n_bits["op_weights"]
+        weight_options = QuantizationOptions(n_bits=weights_n_bits, is_signed=True)
+
+        # Quantize the _X_fit and store the associated quantizer
+        # Weights in KNN algorithms are the train data points
+        # pylint: disable-next=protected-access
+        _X_fit = self.sklearn_model._fit_X
+        q_X_fit = QuantizedArray(
+            n_bits=n_bits["op_weights"],
+            values=numpy.expand_dims(_X_fit, axis=1) if len(_X_fit.shape) == 1 else _X_fit,
+            options=weight_options,
+        )
+        self._q_X_fit = q_X_fit.qvalues
+        self._q_X_fit_quantizer = self._weight_quantizer = q_X_fit.quantizer
+
+        # mypy
+        assert self._q_X_fit_quantizer.scale is not None
+
+        # We assume that the query has the same distribution as the data in _X_fit.
+        # therefore, they use the same scaling and zero point.
+        # https://arxiv.org/abs/1712.05877
+
+        self.output_quant_params = UniformQuantizationParameters(
+            scale=self._q_X_fit_quantizer.scale,
+            zero_point=self._q_X_fit_quantizer.zero_point,
+            offset=0,
+        )
+
+        output_quantizer = UniformQuantizer(params=self.output_quant_params, no_clipping=True)
+
+        assert output_quantizer.zero_point is not None
+        self.output_quantizers.append(output_quantizer)
+
+        # Updating post-processing parameters
+        self._set_post_processing_params()
+
+        self._is_fitted = True
+
+        return self
+
+    def quantize_input(self, X: numpy.ndarray) -> numpy.ndarray:
+        self.check_model_is_fitted()
+        q_X = self.input_quantizers[0].quant(X)
+
+        assert q_X.dtype == numpy.int64, "Inputs were not quantized to int64 values"
+        return q_X
+
+    def dequantize_output(self, q_y_preds: numpy.ndarray) -> numpy.ndarray:
+        self.check_model_is_fitted()
+
+        # We compute the sorted argmax in FHE, which are integers.
+        # No need to de-quantize the output values
+
+        assert q_y_preds[0].shape[-1] == self.n_neighbors, (
```

Comment:
better to check `q_y_preds.n_dim == self.n_neighbors` right ? with `q_y_preds[0].shape[-1]` , having `q_y_preds[0]` empty will fail with wrong message and a shape of `(1, 1, 1, self.n_neighbors)` will wrongly succeed

Reply:
This configuration is not possible because the input must be 2D.
q_y_preds is tuple of size 2, so q_y_preds.ndim returns 2.

Reply:
yes indeed what I said is wrong, but then also, your print message indicates that the predictions must have a shape of `({self.n_neighbors},)` so it's 1d or 2d ? and i any case are you sure it's always going to be 2d ? with neighbors the first dim  ? in any case maybe the assert is not useful enough (you could have let `dequantize_output` just return q_y_preds) !

Reply:
It's not big test, I added it when I had a mismatch dimension due to the extra dim with tuple.
You  even mentioned it in `this does not look right, the output of a circuit is a tuple of one element (since we only pass batches of 1 as inputs), this is why we have this [0]. I can't see why a specific model should do that differently`.

I'll remove it.